src/txn/txn_timestamp.c

```diff
@@ -1056,98 +1122,207 @@ __wt_txn_publish_commit_timestamp(WT_SESSION_IMPL *session)
 			++walked;
 			/*
 			 * Stop on the first entry that we cannot clear.
 			 */
-			if (!qtxn->clear_commit_q)
+			if (!qtxn->clear_durable_q)
 				break;
 
-			TAILQ_REMOVE(&txn_global->commit_timestamph,
-			    qtxn, commit_timestampq);
-			WT_PUBLISH(qtxn->clear_commit_q, false);
-			--txn_global->commit_timestampq_len;
+			TAILQ_REMOVE(&txn_global->durable_timestamph,
+			    qtxn, durable_timestampq);
+			WT_PUBLISH(qtxn->clear_durable_q, false);
+			--txn_global->durable_timestampq_len;
 		}
 
 		/*
 		 * Now walk backwards from the end to find the correct position
 		 * for the insert.
 		 */
 		qtxn = TAILQ_LAST(
-		     &txn_global->commit_timestamph, __wt_txn_cts_qh);
-		while (qtxn != NULL && qtxn->first_commit_timestamp > ts) {
+		     &txn_global->durable_timestamph, __wt_txn_dts_qh);
+		while (qtxn != NULL && __txn_get_durable_timestamp(qtxn) > ts) {
 			++walked;
 			qtxn = TAILQ_PREV(
-			    qtxn, __wt_txn_cts_qh, commit_timestampq);
+			    qtxn, __wt_txn_dts_qh, durable_timestampq);
 		}
 		if (qtxn == NULL) {
-			TAILQ_INSERT_HEAD(&txn_global->commit_timestamph,
-			    txn, commit_timestampq);
-			WT_STAT_CONN_INCR(session, txn_commit_queue_head);
+			TAILQ_INSERT_HEAD(&txn_global->durable_timestamph,
+			    txn, durable_timestampq);
+			WT_STAT_CONN_INCR(session, txn_durable_queue_head);
 		} else
-			TAILQ_INSERT_AFTER(&txn_global->commit_timestamph,
-			    qtxn, txn, commit_timestampq);
-		WT_STAT_CONN_INCRV(session, txn_commit_queue_walked, walked);
+			TAILQ_INSERT_AFTER(&txn_global->durable_timestamph,
+			    qtxn, txn, durable_timestampq);
+		WT_STAT_CONN_INCRV(session, txn_durable_queue_walked, walked);
 	}
-	txn->first_commit_timestamp = ts;
-	++txn_global->commit_timestampq_len;
-	WT_STAT_CONN_INCR(session, txn_commit_queue_inserts);
-	txn->clear_commit_q = false;
-	F_SET(txn, WT_TXN_HAS_TS_COMMIT | WT_TXN_PUBLIC_TS_COMMIT);
-	__wt_writeunlock(session, &txn_global->commit_timestamp_rwlock);
+	++txn_global->durable_timestampq_len;
+	WT_STAT_CONN_INCR(session, txn_durable_queue_inserts);
+	txn->clear_durable_q = false;
+
+	/*
+	 * Mark it has having a PUBLIC durable timestamp so we know to clear it
+	 * out from the queue. But don't mark it as having a durable timestamp
+	 * since we don't want to try to use txn->durable_timestamp or stop
+	 * someone from specifying a durable timestamp later on.
+	 */
+	F_SET(txn, WT_TXN_PUBLIC_TS_DURABLE);
+	__wt_writeunlock(session, &txn_global->durable_timestamp_rwlock);
+}
+
+/*
+ * __wt_txn_publish_durable_timestamp --
+ *	Publish a transaction's durable timestamp.
+ */
+void
+__wt_txn_publish_durable_timestamp(WT_SESSION_IMPL *session)
+{
+	WT_TXN *qtxn, *txn, *txn_tmp;
+	WT_TXN_GLOBAL *txn_global;
+	wt_timestamp_t ts;
+	uint64_t walked;
+
+	txn = &session->txn;
+	txn_global = &S2C(session)->txn_global;
+
+	if (F_ISSET(txn, WT_TXN_PUBLIC_TS_DURABLE))
+		return;
+
+	/*
+	 * Copy the current durable timestamp (which can change while the
+	 * transaction is running).
```

Comment:
Is that true? I'm surprised that another thread could change the durable timestamp for a transaction. I guess it could when it's being cleared from the durable queue - but I doubt that should race with this publish?

Reply:
3e0eab0

That was a leftover from when I copied from the `commit` variant of this function.
Since I've merged the two, this doesn't exist anymore.